### PR TITLE
MM-16809: Added EnsureChannel Helper

### PR DESCRIPTION
**Pull request by ankitrgadiya**
#### Summary

This PR adds `EnsureChannel` helper analogous to `EnsureBot` helper which creates a new channel if not exists and update the meta data of the channel if exists. Also, it will throw error
if the type of existing channel do not match the type of new channel.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/11845
JIRA: https://mattermost.atlassian.net/browse/MM-16809

### CODE REVIEW

File: plugin/helpers_channels.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+	// Check if channel already exists (ignore deleted channels)
+	if existingChannel, chanGetErr := p.API.GetChannelByName(channel.TeamId, channel.Name, false); chanGetErr == nil && existingChannel != nil {
+		// Update metadata of the channel
+		if updateErr := updateChannelMeta(existingChannel, channel); updateErr != nil {
+			return "", errors.Wrap(updateErr, "Failed to update the metadata of existing channel")
+		}
+
+		// Send the updates to API
+		updatedChannel, chanUpdateErr := p.API.UpdateChannel(existingChannel)
```
Comment (ankitrgadiya) [inline plugin/helpers_channels.go]: One minor optimization here can be that I only call `UpdateChannel` function if the metadata was actually updated.

**Comment by ankitrgadiya** [comment]
@lieut-data Can you please assign reviewers on this PR. Thanks

File: plugin/helpers_channels.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+func (p *HelpersImpl) EnsureChannel(channel *model.Channel) (retChanId string, retErr error) {
+	// Must provide a channel with a name and teadId
+	if channel == nil || len(channel.Name) < 1 || len(channel.TeamId) < 1 {
+		return "", errors.New("passed a bad channel, nil or no name or no team id")
+	}
+
+	// If we fail for any reason, this could be a race between creation of channel and
+	// retrieval from another EnsureChannel. Just try the basic retrieve existing again.
+	defer func() {
+		if retChanId == "" || retErr != nil {
```
[the remark this answers is not in the document]
Comment (ankitrgadiya) [reply plugin/helpers_channels.go]: Yep, makes sense. I'll replace it.

File: plugin/helpers_channels.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	// Check if Channel ID exists
+	if chanIdBytes != nil {
+		chanId := string(chanIdBytes)
+		return chanId, nil
+	}
+
+	// Check if channel already exists (ignore deleted channels)
+	if existingChannel, chanGetErr := p.API.GetChannelByName(channel.TeamId, channel.Name, false); chanGetErr == nil && existingChannel != nil {
```
[the remark this answers is not in the document]
Comment (ankitrgadiya) [reply plugin/helpers_channels.go]: Yes, I should check for the error. I'll fix it, thanks :)

File: plugin/helpers_channels_test.go [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+package plugin_test
+
+import (
+	"testing"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/plugin"
+	"github.com/mattermost/mattermost-server/plugin/plugintest"
+	// "github.com/mattermost/mattermost-server/plugin/plugintest/mock"
```
[the remark this answers is not in the document]
Comment (ankitrgadiya) [reply plugin/helpers_channels_test.go]: I'll remove it :)

**Comment by ankitrgadiya** [comment]
@marianunez I made the changes, can you please review the changes?

File: plugin/helpers_channels.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	channelIdBytes, kvGetErr := p.API.KVGet(CHANNEL_KEY)
+	if kvGetErr != nil {
+		// Failed to retrive the value of channel
+		return "", errors.Wrap(kvGetErr, "failed to get channel ID")
+	}
+
+	// Check if Channel ID exists
+	if channelIdBytes != nil {
+		channelId := string(channelIdBytes)
+		return channelId, nil
```
Comment (crspeller) [inline plugin/helpers_channels.go]: We don't update the metadata if the channel already exists?

File: plugin/helpers_channels.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	// Check if channel already exists (ignore deleted channels)
+	existingChannel, channelGetErr := p.API.GetChannelByName(channel.TeamId, channel.Name, false)
+	if channelGetErr != nil {
+		return "", errors.Wrap(channelGetErr, "failed to get channel")
+	} else if channelGetErr == nil && existingChannel != nil {
```
Comment (crspeller) [inline plugin/helpers_channels.go]: Since this is in an else `channelGetErr == nil` is redundant.

File: plugin/helpers_channels.go (hunk 5 above)
Comment (ankitrgadiya) [reply plugin/helpers_channels.go]: Yes It should update the metadata if channel exists. I will fix it.

File: plugin/helpers_channels.go (hunk 6 above)
Comment (ankitrgadiya) [reply plugin/helpers_channels.go]: I will remove it. Thanks :)

File: plugin/helpers_channels.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	// Channel exists!
+	return updatedChannel.Id, nil
+}
+
+func updateChannelMeta(existing *model.Channel, new *model.Channel) error {
+	// Check if channels are of different types
+	if existing.Type != new.Type {
+		return errors.New("Channel type cannot be updated")
```
Comment (crspeller) [inline plugin/helpers_channels.go]: Why not?

Comment (ankitrgadiya) [reply plugin/helpers_channels.go]: Not sure but it was specified in the issue

Comment (crspeller) [reply plugin/helpers_channels.go]: "It might not be safe to automatically fix this." True, if we stole a random channel by name that might not be great.

**Comment by hanzei** [comment]
Thank you very much for the contribution @ankitrgadiya :tada:

**Pull request merged**